packages/test/test-utils/src/loaderContainerTracker.ts

```diff
@@ -25,36 +54,470 @@ export class LoaderContainerTracker {
         loader.rehydrateDetachedContainerFromSnapshot = patch(loader.rehydrateDetachedContainerFromSnapshot);
     }
 
+    /**
+     * Utility function to add container to be tracked.
+     *
+     * @param container - container to add
+     */
+    private addContainer(container: IContainer) {
+        // ignore summarizer
+        if (!container.deltaManager.clientDetails.capabilities.interactive) { return; }
+
+        // don't add container that is already tracked
+        if (this.containers.has(container)) { return; }
+
+        const record = {
+            index: this.containers.size,
+            paused: false,
+            startTrailingNoOps: 0,
+            trailingNoOps: 0,
+            lastProposal: 0,
+        };
+        this.containers.set(container, record);
+        this.trackTrailingNoOps(container, record);
+        this.trackLastProposal(container);
+        this.setupTrace(container, record.index);
+    }
+
+    /**
+     * Keep track of the trailing NoOp that was sent so we can discount them in the clientSequenceNumber tracking.
+     * The server might coalesce them with other ops, or a single NoOp, or delay it if it don't think it is necessary.
+     *
+     * @param container - the container to track
+     * @param record - the record to update the trailing op information
+     */
+    private trackTrailingNoOps(container: IContainer, record: ContainerRecord) {
+        container.deltaManager.outbound.on("op", (messages) => {
+            for (const msg of messages) {
+                if (msg.type === MessageType.NoOp) {
+                    // Track the NoOp that was sent.
+                    if (record.startTrailingNoOps === 0) {
+                        record.startTrailingNoOps = msg.clientSequenceNumber;
+                    }
+                    record.trailingNoOps++;
+                } else {
+                    // Other ops has been sent. We would like to see those ack'ed, so no more need to track NoOps
+                    record.startTrailingNoOps = 0;
+                    record.trailingNoOps = 0;
+                }
+            }
+        });
+
+        container.deltaManager.inbound.on("push", (message) => {
+            // Received the no op back, update the record.
+            if (message.type === MessageType.NoOp
+                && message.clientId === (container as Container).clientId
```

Comment:
I really dislike how we always need to cast to container. we should consider #1707

Reply:
I am trying not to lump that into this PR.